platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -1142,48 +1562,149 @@ namespace platoon_strategic_ihp
+                return MobilityRequestResponse::ACK;
+            }
+            /*
+            *************** Reserve for cut-in from middle and cut-in from rear. *****************
+            // cut-in rear
+            else if ((join_index == pm_.platoon.size()-1) && isRearJoinerInPosition)
+            {
+                // todo: send request to last member to slow down (i.e., set isCreateGap to true) 
+                request. 
+                request.
+                ROS_DEBUG_STREAM("");
+                return MobilityRequestResponse::ACK;
+            }
+            // cut-in and join in middle
+            else if (isMidJoinerInPosition)
+            {
+                // todo: send request to index member to slow down (i.e., set isCreateGap to true) 
+                request. 
+                request.
+                ROS_DEBUG_STREAM("");
+                return MobilityRequestResponse::ACK;
+            }
+            */
+        }
+
+        // task 2: For cut-in from front, the leader need to stop creating gap when gap is big enough
+        if (plan_type.type == cav_msgs::PlanType::STOP_CREATE_GAP && pm_.isCreateGap) 
+        {
+            // reset creat gap indicator
+            pm_.isCreateGap = false;
+            // no need to response, simple reset the indicator
+            return MobilityRequestResponse::NO_RESPONSE;
+
+        }
 
-        lanelet::BasicPoint2d incoming_pose = ecef_to_map_point(msg.location);
-        double obj_cross_track = wm_->routeTrackPos(incoming_pose).crosstrack;
-        bool inTheSameLane = (abs(obj_cross_track - current_crosstrack_) < config_.maxCrosstrackError);
-        ROS_DEBUG_STREAM("current_cross_track error = " << abs(obj_cross_track - current_crosstrack_));
-        ROS_DEBUG_STREAM("inTheSameLane = " << inTheSameLane);
-        if (isTargetVehicle && isCandidateJoin && inTheSameLane)
+        // task 3.1 cut-in front: After creating gap, revert back to same-lane operation 
+        if (plan_type.type == cav_msgs::PlanType::CUT_IN_FRONT_DONE)
         {
-            ROS_DEBUG_STREAM("New joiner vehicle " << fj_new_joiner_Id_ << " (previous leader) is joining .");
-            ROS_DEBUG_STREAM("Changing to PlatoonLeaderState and send ACK to the new joiner vehicle");
-            pm_.current_platoon_state = PlatoonState::LEADER;
+            ROS_DEBUG_STREAM("Cut-in from front lane change finished, leader revert to same-lane maneuver, leader reset member variable joiningID_.");
+            joiningID_ = "";
+            pm_.current_platoon_state = PlatoonState::LEADERABORTING;
             return MobilityRequestResponse::ACK;
         }
-        else
+
+        // todo: task 3 cut-in from middle/rear
+
+        // task 4: if other joining vehicle send joning request, NACK it since there is already a cut-in join going on.
+        if (plan_type.type == cav_msgs::PlanType::CUT_IN_FROM_SIDE)
```

Comment:
why do you need to specify the CUT_IN_FROM_SIDE? cant you just use the "else" to reject other requests?
What if you get like a join_from_rear req by mistake?

Reply:
Indeed. Updated to else. This should definitely apply to all requests.